### PR TITLE
ARROW-2106: [Python] Add conversion for a series of datetime objects

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/python/numpy_to_arrow.cc [hunk 1]
```diff
@@ -783,6 +784,35 @@ Status NumPyConverter::ConvertDecimals() {
...
 
+Status NumPyConverter::ConvertDateTimes() {
+  // Convert array of datetime.datetime objects to Arrow
+  PyAcquireGIL lock;
+  PyDateTime_IMPORT;
+
+  Ndarray1DIndexer<PyObject*> objects(arr_);
+
+  // datetime.datetime stores microsecond resolution
+  Time64Builder builder(::arrow::timestamp(TimeUnit::MICRO), pool_);
```
Comment (wesm) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: This should be `TimestampBuilder`?

File: cpp/src/arrow/python/numpy_to_arrow.cc [hunk 2]
```diff
@@ -783,6 +784,35 @@ Status NumPyConverter::ConvertDecimals() {
...
+  RETURN_NOT_OK(builder.Resize(length_));
+
+  PyObject* obj = NULLPTR;
```
Comment (wesm) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: NB it is not necessary AFAIK to use `NULLPTR` in compilation units, only in header files

File: cpp/src/arrow/python/numpy_to_arrow.cc [hunk 3]
```diff
@@ -1005,6 +1035,8 @@ Status NumPyConverter::ConvertObjectsInfer() {
     } else if (PyDate_CheckExact(obj)) {
       // We could choose Date32 or Date64
       return ConvertDates<Date32Type>();
+    } else if (PyDateTime_CheckExact(obj)) {
```
Comment (wesm) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: I opened https://issues.apache.org/jira/browse/ARROW-2130 about dealing with pandas.Timestamp which will inevitably pop up

File: python/pyarrow/tests/test_convert_pandas.py [hunk 4]
```diff
@@ -649,6 +649,19 @@ def test_timestamps_with_timezone(self):
 
         _check_pandas_roundtrip(df)
 
+    def test_python_datetime(self):
+        # ARROW-2106
+        date_array = [datetime.today() + timedelta(days=x) for x in range(10)]
+        df = pd.DataFrame({
+            'datetime': pd.Series(date_array, dtype=object)
+        })
+        result = pa.Table.from_pandas(df).to_pandas()
```
Comment (wesm) [inline python/pyarrow/tests/test_convert_pandas.py]: Can you add an assertion that the column in the result of `pyarrow.Table.from_pandas` is `TimestampArray` (per the comment above). I expect that is already the case

**Review by wesm — COMMENTED** [review]
Thanks for doing this, a couple of minor comments. I may go ahead and add the couple tweaks quickly

**Comment by wesm** [comment]
+1, will merge on passing build.

File: cpp/src/arrow/python/numpy_to_arrow.cc (hunk 3 above)
Comment (xhochy) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: I'll take care of that in a follow-up PR.

File: cpp/src/arrow/python/numpy_to_arrow.cc (hunk 1 above)
Comment (xhochy) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: Yes, this is a copy&paste artefact. Probably some code duplication we may eliminate in the future.

**Pull request closed without merging**